client/engine/store/mockstore.go

```diff
@@ -50,13 +53,48 @@ func (ms MockStore) GetChannelSecretKey() *[]byte {
 
 func (ms MockStore) GetObjectiveById(id protocols.ObjectiveId) (obj protocols.Objective, ok bool) {
 	// todo: locking
-	obj, ok = ms.objectives[id]
+	objJSON, ok := ms.objectives[id]
+
+	if strings.HasPrefix(string(id), "DirectFunding-") {
+		// unmarshal directFundObjective
+		var jsonDFO directfund.JSONDirectFundObjective
+		json.Unmarshal([]byte(objJSON), &jsonDFO)
+		channel, _ := ms.getChannelById(jsonDFO.C)
+		jsonDFO.DirectFundObjective.C = &channel
+
+		obj = jsonDFO.DirectFundObjective
+	} else if strings.HasPrefix(string(id), "VirtualFund-") {
+		// unmarshal virtalFundObj
+		var jsonVFO virtualfund.JSONVirtualFundObjective
+		json.Unmarshal([]byte(objJSON), &jsonVFO)
+		channel, _ := ms.getChannelById(jsonVFO.V)
+		jsonVFO.VirtualFundObjective.V = &channel
+
+		obj = jsonVFO.VirtualFundObjective
+	}
+
 	return obj, ok
 }
 
 func (ms MockStore) SetObjective(obj protocols.Objective) error {
 	// todo: locking
-	ms.objectives[obj.Id()] = obj
+	bytes, err := obj.MarshalJSON()
+	if err != nil {
+		return err
+	}
+	ms.objectives[obj.Id()] = string(bytes)
+
+	id := obj.Id()
+	if strings.HasPrefix(string(id), "DirectFunding-") {
```

Comment:
This reminds me of a "discriminated union" which is a pattern from Typescript. Elsewhere we have managed to pivot to using the "visitor pattern" instead. Did you consider that option? What would be the pros and cons?

More specifically, could you simply call a new interface method `obj.getChannels` to get an array of channels to pass to `ms.setChannel`? Then, the store does not need to know what kind of objective it is dealing with, only that it can trust it to report which channels need persisting or retrieving.